firewalldb/actions_sql.go

```diff
@@ -0,0 +1,418 @@
+func (s *SQLDB) AddAction(ctx context.Context,
+	req *AddActionReq) (ActionLocator, error) {
+
+	var (
+		writeTxOpts db.QueriesTxOptions
+		locator     sqlActionLocator
+
+		actor = sql.NullString{
+			String: req.ActorName,
+			Valid:  req.ActorName != "",
+		}
+		feature = sql.NullString{
+			String: req.FeatureName,
+			Valid:  req.FeatureName != "",
+		}
+		trigger = sql.NullString{
+			String: req.Trigger,
+			Valid:  req.Trigger != "",
+		}
+		intent = sql.NullString{
+			String: req.Intent,
+			Valid:  req.Intent != "",
+		}
+	)
+
+	err := s.db.ExecTx(ctx, &writeTxOpts, func(db SQLQueries) error {
+		// Do best effort to see if this action is linked to a session,
+		// and/or an action or none.
+		var (
+			sessionID sql.NullInt64
+			accountID sql.NullInt64
+		)
+
+		// First check session DB.
+		var sessErr error
+		req.SessionID.WhenSome(func(alias session.ID) {
+			sessID, err := db.GetSessionIDByAlias(ctx, alias[:])
+			if errors.Is(err, sql.ErrNoRows) {
+				sessErr = session.ErrSessionNotFound
+				return
+			} else if err != nil {
+				sessErr = err
+				return
+			}
+
+			sessionID = sqldb.SQLInt64(sessID)
+		})
+		if sessErr != nil {
+			return sessErr
+		}
+
+		// If an account ID was provided, then it must exist in our DB.
+		var getAcctErr error
+		req.AccountID.WhenSome(func(alias accounts.AccountID) {
+			aliasInt, err := alias.ToInt64()
+			if err != nil {
+				getAcctErr = err
+				return
+			}
+
+			acctID, err := db.GetAccountIDByAlias(ctx, aliasInt)
+			if errors.Is(err, sql.ErrNoRows) {
+				getAcctErr = accounts.ErrAccNotFound
+				return
+			} else if err != nil {
+				getAcctErr = err
+				return
+			}
+
+			accountID = sqldb.SQLInt64(acctID)
+		})
+		if getAcctErr != nil {
+			return getAcctErr
+		}
+
+		var macID []byte
+		req.MacaroonIdentifier.WhenSome(func(id [4]byte) {
+			macID = id[:]
+		})
+
+		id, err := db.InsertAction(ctx, sqlc.InsertActionParams{
+			SessionID:          sessionID,
+			AccountID:          accountID,
+			ActorName:          actor,
+			MacaroonIdentifier: macID,
+			FeatureName:        feature,
+			ActionTrigger:      trigger,
+			Intent:             intent,
+			StructuredJsonData: []byte(req.StructuredJsonData),
+			RpcMethod:          req.RPCMethod,
+			RpcParamsJson:      req.RPCParamsJson,
+			CreatedAt:          s.clock.Now().UTC(),
+			ActionState:        int16(ActionStateInit),
+		})
+		if err != nil {
+			return err
+		}
+
+		locator = sqlActionLocator{
+			id: id,
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &locator, nil
+}
+
+// SetActionState finds the action specified by the ActionLocator and sets its
+// state to the given state.
+//
+// NOTE: This is a part of the ActionDB interface.
+func (s *SQLDB) SetActionState(ctx context.Context, al ActionLocator,
+	state ActionState, errReason string) error {
+
+	if errReason != "" && state != ActionStateError {
+		return fmt.Errorf("error reason should only be set for " +
+			"ActionStateError")
+	}
```

Comment:
Nit: As a follow-up to @bitromortac's feedback below, we could potentially also check that an `errReason` is actually set when `state == ActionStateError`, though that may be a bit dangerous to require that the middleware interceptor sets an error for all responses marked as "errors". So therefore I think it's probably safer to leave it as is.